feat(flow): add auto arrange nodes to canvas controls (#8295)

Aayush-Gangwar opened the pull request:
## Summary

This PR introduces a new **Organize Nodes** feature in the Langflow canvas controls.
Users can now automatically organize nodes on the canvas in a clean, logical sequence, improving readability and reducing visual clutter for complex flows.

## Motivation

As noted in Issue #8295, flows can become messy when multiple nodes are added.
This feature provides a **one-click solution** to arrange nodes and their edges automatically, enhancing usability and workflow efficiency.

### Changes

* Added an **"Organize Nodes" button** in the canvas controls toolbar.
* Utilized the **existing ELK.js auto-layout** in Langflow to automatically arrange nodes and edges.
* Updated state management to support auto-arrangement.
* Snapshots are recorded for undo/redo functionality.
* Disabled the Auto Arrange button when the flow is locked.

## Commit breakdown
- **Commit 1:** Adds the auto-arrange UI and functionality using default layout behavior
- **Commit 2:** Proposes improved layout options (spacing, node placement, stability) for improved visuals; maintainers can review this commit independently.


## Screenshots
### Before Auto-Arrange
Messy layout after adding multiple components manually.
<img width="1916" height="870" alt="Screenshot 2025-12-14 154242" src="https://github.com/user-attachments/assets/5f0e4f4b-1ad3-4ea1-9e3f-0a217113df16" />

### After Auto-Arrange
Nodes organized left-to-right with improved spacing and readability.
<img width="1918" height="866" alt="Screenshot 2025-12-14 154733" src="https://github.com/user-attachments/assets/d9971f4b-3e54-49a4-822e-68e8da431e46" />


## Related Issue

Closes #8295

## Notes
* Shortcut key support (e.g., Ctrl+Arrange) may be added in future iterations.
* FitView behavior is left to user’s default; auto-viewport adjustments can be added later.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added "Organize Nodes" button to canvas controls for automatically arranging and positioning all nodes in your flow diagram (button is disabled when the flow is locked).
  * Enhanced layout algorithm with improved node and edge spacing to create cleaner, more organized flow diagrams with better visual clarity.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Aayush-Gangwar committed:
feat(flow): add auto arrange nodes functionality and button to canvas controls (#8295)

Aayush-Gangwar committed:
chore(flow): suggest improved layout options for nodes placement

Aayush-Gangwar commented:
Hi @AntonioABLima 👋

Just wanted to gently follow up on this PR. It’s been a while since the last update, and I wanted to check if it’s in a good state for review.

This adds the "Organize Nodes" feature using ELK.js for auto-layout — happy to make any changes if needed.

Thanks for your time 🙂

Prgckwb commented:
+1,  I independently built the same feature and found this PR. Would love to see it land. Is there anything blocking the review?

Aayush-Gangwar commented:
Rebased this onto current `main` and reworked the implementation — canvas controls got a significant refactor upstream since this PR was opened (new assistant panel, minimize-all hook, dropdown-based layout), so the original diff no longer applied cleanly on top of it.

What changed from the previous version:
- Moved arrange logic into a dedicated `useAutoArrange` hook (`canvasControlsComponent/hooks/use-auto-arrange.ts`), mirroring the existing `use-minimize-all-and-align` hook, instead of patching `MemoizedCanvasControls` directly.
- Fixed a race condition in the original version: it replaced the whole node list with the async layout result, which could discard any node added/edited while ELK.js was still computing. Positions are now merged by id into the current node list, with guards against the canvas unmounting or the active flow changing mid-computation.
- Added i18n keys instead of hardcoded strings.
- Added test coverage (`useAutoArrange.test.tsx`) — this addresses the earlier test-coverage CI failure.
- Dropped the ELK `layoutOptions` tuning from the previous version's second commit — that constant is shared with the existing import-time auto-layout path (`reactflowUtils.ts`), so changing it here would've silently altered unrelated behavior too. Open to revisiting spacing as a separate, focused PR if there's interest.

Squashed into a single commit since this is a rewrite rather than an incremental fix on the old one. Happy to split differently if preferred.

Aayush-Gangwar commented:
Both of the failing checks look unrelated to this PR — the **Docker image build** failed on a plain `npm error network read ECONNRESET` while installing deps, and the **Playwright shard 47/70** failure is on `groupFalseError.spec.ts`, which has nothing to do with canvas controls (looks like a backend timing issue, not a frontend regression — the shard
even listed different tests on the retry vs. the first run).

Could someone with access re-run the failed jobs? Happy to look closer if they fail again. 
Thanks!
